### ISSUE TITLE
End-to-end automation pretty much broken

### ISSUE THREAD

**Issue by TarasRoshko** [Azure/service-fabric-issues#131]
Objective: deploy a new SF cluster and immediately deploy an app to it, all fully automatic using PS cmdlets
Reality: `New-AzureRmResourceGroupDeployment` with SF template finishes when SF cluster is still in "Deploying" state so subsequent `Connect-ServiceFabricCluster` calls will just fail. This can be worked around by looping `Connect-ServiceFabricCluster` until it succeeds. Next issue is, it succeeds when cluster itself is still "Deploying" according to the portal, so trying to `Publish-NewServiceFabricApplication` after successfully `Connect-ServiceFabricCluster` fails 100% of the time, it requires some other polling which I can't figure out at the moment (can't find a cmdlet returning cluster state as shown on portal).

So right now the automation is:

- Deploy template
- Loop `Connect-ServiceFabricCluster` *[see update]
- Poll on (idk what) until cluster is "Ready" *[see update]
- Deploy and app using `Publish-NewServiceFabricApplication`

The whole point of PS API is to enable end-to-end automation which is not possible without multiple workarounds. 

The API should provide an option to wait until cluster is available OR return a future so the caller can wait on it if necessary. Similar to https://github.com/Azure/azure-powershell/issues/2979

**UPDATE 1**

So instead of looping `Connect-ServiceFabricCluster`, one can do this:

```
    while (($state = (Get-AzureRmResource -ResourceGroupName $resourceGroupName -Name $clusterName -ResourceType "Microsoft.ServiceFabric/clusters").Properties.clusterState) -ne "Ready")
    {
        Log::Info "'$clusterName' is '$state', retrying in 32s"
        Start-Sleep -Seconds 32
    }

    Log::Info "Connecting to '$url'"

    Connect-ServiceFabricCluster -ConnectionEndpoint $url `
                                 -X509Credential `
                                 -ServerCertThumbprint $sfCommunicationCertThumbprint `
                                 -FindType FindByThumbprint `
                                 -FindValue $sfDeploymentCertThumbprint `
                                 -StoreLocation "CurrentUser" `
                                 -StoreName "My" | Out-Null
```

And even this does not seem to be good enough: 

```
[deploy.ps1:220 I] Connecting to Service Fabric Cluster 'appstrmsfcontrollertaras' at 'appstrmsfcontrollertaras.westus.cloudapp.azure.com:19000'
[deploy.ps1:593 I] 'appstrmsfcontrollertaras' is 'Deploying', retrying in 16s
/30m of waiting/
[deploy.ps1:597 I] Connecting to 'appstrmsfcontrollertaras.westus.cloudapp.azure.com:19000'
WARNING: Cluster connection with the same name already existed, the old connection will be deleted
[deploy.ps1:233 I] Deploying app 'fabric:/controller' to Service Fabric Cluster at 'appstrmsfcontrollertaras.westus.cloudapp.azure.com:19000'
WARNING: Failed to contact Naming Service. Attempting to contact Failover Manager Service...
WARNING: Failed to contact Failover Manager Service, Attempting to contact FMM...
WARNING: No such host is known
WARNING: Unable to Verify connection to Service Fabric cluster.
[common.psm1:30 E] 'Deploy' failed. Details:
------------------------------------------------------------------------------------------------------------------------
System.Fabric.FabricException: No cluster endpoint is reachable, please check if there is connectivity/firewall/DNS issue.
   at System.Management.Automation.MshCommandRuntime.ThrowTerminatingError(ErrorRecord errorRecord)
------------------------------------------------------------------------------------------------------------------------
```

So status was "Ready", `Connect-ServiceFabricCluster` succeed, and yet publishing failed. Tried manually 5m later - failed. Couple more minutes - succeeded.

At this point I'm considering looping `Publish-NewServiceFabricApplication`, stay tuned

**Update 2**

So the resource polling, connecting and publishing seems to be enough, that was another issue with unavailable `$clusterConnection` variable (see #137)

Overall, something like this is required:

```

        while (($state = (Get-AzureRmResource -ResourceGroupName $resourceGroupName -Name $clusterName -ResourceType "Microsoft.ServiceFabric/clusters").Properties.clusterState) -ne "Ready")
        {
            Log::Info "'$clusterName' is '$state', retrying in 32s"
            Start-Sleep -Seconds 32
        }

        Connect-ServiceFabricCluster -ConnectionEndpoint $clusterEndpoint `
                                     -X509Credential `
                                     -ServerCertThumbprint $sfCommunicationCertThumbprint `
                                     -FindType FindByThumbprint `
                                     -FindValue $sfDeploymentCertThumbprint `
                                     -StoreLocation "CurrentUser" `
                                     -StoreName "My"

        $global:clusterConnection = $clusterConnection

        return Publish-NewServiceFabricApplication -ApplicationName $name `
                                                   -Action "RegisterAndCreate" `
                                                   -OverwriteBehavior "Always" `
                                                   -ApplicationPackagePath $package `
                                                   -ApplicationParameter $settings
```

**Comment by MikkelHegn** [comment]
Looking forward to more updates from you Taras.

If you stay on the path, that tooling and the SDK gives you - you are able to deploy fully automated using VSTS. This article describes that end-to-end scenario: https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-set-up-continuous-integration

If you need more control over the individual steps, you can either work through PS as you are doing above, but there are some limitations as you have discovered. Another approach is to for the deployment tasks off the vsts-tasks repo and deploy your own version to your VSTS instance. https://github.com/Microsoft/vsts-tasks/tree/master/Tasks/ServiceFabricDeploy

**Comment by TarasRoshko** [comment]
We have `deploy.ps1` which is single script to deploy entire solution to Azure, it consists of many more resources and Service Fabric is just one of them. It is used by developers to deploy personal environment, we also use it to deploy `dev` `prod` `test` and so on, we already have shared parts extracted into VSTS tasks (e.g. Windows Image Generation). So unfortunately things like this just cannot go into VSTS.

Think developer deploying his own testing environment which has to be exactly the same as prod, just scaled down a lot. This is what our `deploy.ps1` does - deploys entire environment

**Comment by masnider** [comment]
@TarasRoshko - the new Azure PS cmds (ex: https://docs.microsoft.com/en-us/powershell/module/azurerm.servicefabric/new-azurermservicefabriccluster?view=azurermps-4.3.1) should make this easier for you from the SF perspective. Since that only completes when the cmdlet can connect to the cluster. I think that addresses the main point of your issue here, so I'm going to close down this issue. If you have any other questions or issues/feedback with that cmdlet, just create a new issue and we can handle it. Thanks for the feedback!
